### PR TITLE
feat(*): add `OmitKeyof` to Omit object by object's keys strictly

### PR DESCRIPTION
**Pull request by manudeli**
This change add `OmitKeyof` to validate Omitting keys should be keyof Omitted Object Type

As shown below, the inconvenience of having to manually remove unnecessary fields can be identified early on at the type level with `OmitKeyof`.

### Motivatings
- https://github.com/TanStack/query/pull/6907/files
- https://github.com/TanStack/query/pull/7139#discussion_r1530890783

### Expectation
I made test-d to type-test `OmitKeyof` in https://github.com/TanStack/query/pull/7139/files#diff-e6421be8af92bef480137fd70c8e935fd1975d73ae3a75f393893de5542e2a41
- For complex types using `Omit`, I want to reduce the difficulty of library maintenance for it.
    - `OmitKeyof` provide autocomplete by extending keyof TObject
       Mar-22-2024 18-12-27
    - `OmitKeyof` check strictly second type parameter of it by extending keyof TObject
       Mar-22-2024 18-13-45
    - `OmitKeyof` prevent library contributors make misspell like queryTey or qeuryKey
       Mar-22-2024 18-14-35
- Library users can be supplied with stable types consistently

### CODE REVIEW

File: packages/react-query/src/useQueries.ts [hunk 1]
```diff
@@ -43,9 +44,9 @@ type UseQueryOptionsForUseQueries<
   TError = DefaultError,
   TData = TQueryFnData,
   TQueryKey extends QueryKey = QueryKey,
-> = Omit<
+> = OmitKeyOf<
   UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
-  'placeholderData' | 'suspense'
```
Comment (manudeli) [inline packages/react-query/src/useQueries.ts:48]: This unnecessary suspense field could be checked on type level by OmitKeyOf

File: packages/react-query/src/__tests__/useQueries.test-d.tsx [hunk 2]
```diff
@@ -104,7 +105,7 @@ describe('UseQueries config object overload', () => {
       type Data = string
 
       const useCustomQueries = (
-        options?: Omit<UseQueryOptions<Data>, 'queryKey' | 'queryFn'>,
+        options?: OmitKeyOf<UseQueryOptions<Data>, 'queryKey' | 'queryFn'>,
```
Comment (manudeli) [inline packages/react-query/src/__tests__/useQueries.test-d.tsx:108]: We don't have to worry our misspell like below

```ts
type NoNecessaryError = Omit<UseQueryOptions<Data>, 'queryTey' | 'qeuryFn'>,
```

queryTey, qeuryFn(wrong spell) make type error if with OmitKeyOf

```ts
type NecessaryError = OmitKeyOf<UseQueryOptions<Data>, 'queryTey' | 'qeuryFn'>, // will make type error
```

**Review by manudeli — COMMENTED** [review]
I left my intentions

File: packages/vue-query/src/__tests__/useQueries.types.test.ts [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import { describe, it } from 'vitest'
-import { reactive } from 'vue'
-import { useQueries } from '..'
+import { UnwrapRef, reactive } from 'vue'
```
Comment (manudeli) [inline packages/vue-query/src/__tests__/useQueries.types.test.ts:2]:
```suggestion
import { reactive } from 'vue'
```

File: packages/vue-query/src/__tests__/useQueries.types.test.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import { describe, it } from 'vitest'
 import { reactive } from 'vue'
-import { useQueries } from '..'
+import { OmitKeyOf, useQueries } from '..'
```
Comment (manudeli) [inline packages/vue-query/src/__tests__/useQueries.types.test.ts:3]:
```suggestion
import { useQueries } from '..'
```

**Review by TkDodo — APPROVED** [review]

**Pull request merged**